File: defmi/netting.py
"""Netting cycles: gross-gross, gross-net, net-net, without reading anything.

The three shapes are the BIS delivery-versus-payment models --- model 1 settles
both legs trade by trade, model 2 settles securities gross and cash net, model 3
nets both. Here the choice is not only a settlement-system question: it decides
how many range proofs exist and when, which is the one cost in this design that
hurts.

A rail is either gross or net, and that single distinction carries everything.

A **gross** rail checks each order as it arrives. The participant proves that its
position after the order is still non-negative, so an order that would leave it
short is refused before it exists and settlement failure cannot happen. The price
is that admission is order-sensitive: a participant due to receive a hundred and
deliver a hundred is refused if the delivery arrives first, which is exactly the
liquidity benefit netting exists to provide, given up on purpose.

A **net** rail accumulates. Positions move homomorphically with no proof at all,
and because a commitment hides a value's sign as readily as its magnitude, an
intermediate position may be negative without anything leaking --- nothing is
proved about it, so nothing about it is disclosed. One range proof per
participant at the close establishes that the *net* is covered. That restores the
liquidity benefit and makes admission order-insensitive, at the cost of a cycle
that can fail at the close, which is the classic risk stated in its usual place.

Neither arrangement needs the offset trick that hiding a signed position would
otherwise require: on a gross rail nothing is ever negative, and on a net rail
nothing intermediate is ever proved.

There is a second, larger knob, and measuring the first one is what exposed it.
Netting the rails saves only the per-order cover proof, because every order still
carries an instruction the settlement layer has to verify, and that verification
--- which contains its own range proofs on amount and price --- does not net.
What does net is the *instruction*. A cycle can be run against one quorum
attestation over the closing positions instead of one instruction per trade, and
then the settlement layer's work stops depending on the number of trades at all.
The saving is real but it is not free: with a batch attestation the layer no
longer checks the individual trades, so the split between participants is
attested by the quorum rather than verified. Conservation and solvency still hold
outright, and each participant can still check its own net, so what is given up
is third-party verifiability of the split --- the same bargain a central
counterparty represents, made explicit.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from enum import Enum
from typing import Any

from zk.commit import (
    CrossGeneratorProof, Pedersen, ProductProof, RangeProof, prove_bounded,
    prove_product, prove_range, prove_same_value, verify_bounded, verify_product,
    verify_range, verify_same_value,
)
from .credit import CreditLine, check_credit
from zk.groups import Group
from zk.zkpi import SettlementVenue, ZkPaymentInstruction

CYCLE_DOMAIN = b"QOMM:DEFMI:CYCLE:v1"
MAX_POSITION = (1 << 40) - 1


class NettingMode(Enum):
    """Which rails defer their coverage check to the close of the cycle."""

    GROSS_GROSS = "gross-gross"      # BIS DVP model 1
    GROSS_NET = "gross-net"          # model 2: securities gross, cash net
    NET_NET = "net-net"              # model 3

    @property
    def securities_net(self) -> bool:
        return self is NettingMode.NET_NET

    @property
    def cash_net(self) -> bool:
        return self is not NettingMode.GROSS_GROSS


class Rejected(Exception):
    """An order that would leave a gross position short."""


@dataclass
class Position:
    handle: bytes
    commitment: Any
    opening_commitment: Any
    moves: int = 0


@dataclass(frozen=True)
class Leg:
    """One rail's movement: one delta, one payer, one payee.

    Deliberately not a pair of independent debit and credit records. With one
    delta there is no way for the two sides to disagree, so conservation on the
    rail holds by construction rather than by a check that could be forgotten.

    ``cover`` is present only on a gross rail. The payer's new position is never
    sent: the book derives it as old/delta, a group operation, and checks the
    proof against what it derived.
    """

    payer: bytes
    payee: bytes
    delta: Any
    delta_link: CrossGeneratorProof
    cover: RangeProof | None = None


@dataclass(frozen=True)
class Order:
    """One trade's effect on four positions.

    ``cash_reference`` is a commitment to quantity times price under the base
    generator, with ``value_proof`` relating it to the quantity and price the
    quorum signed. The instruction commits to the factors, never to the product,
    so without it the cash leg would have nothing to be checked against.
    """

    instruction: ZkPaymentInstruction
    securities: Leg
    cash: Leg
    cash_reference: Any
    value_proof: ProductProof


@dataclass(frozen=True)
class BatchAttestation:
    """The quorum standing behind a whole cycle instead of each trade in it.

    Signed over the closing position commitments, so it says nothing about any
    individual trade and cannot be replayed onto a different set of positions.
    """

    digest: bytes
    signature: bytes
    quorum: tuple


@dataclass(frozen=True)
class Coverage:
    """A participant showing, at the close, that its net is covered."""

    handle: bytes
    proof: RangeProof


@dataclass(frozen=True)
class CycleReceipt:
    mode: str
    admitted: int
    refused: int
    status: str
    reason: str
    securities_before: str
    securities_after: str
    cash_before: str
    cash_after: str
    closed_at: int


class PositionBook:
    """Positions on one rail, under a tag that is stable for the cycle.

    Stability is required for positions to accumulate at all; the tag is
    refreshed when the cycle rolls, so what a reader can group together is
    confined to one cycle rather than following an asset.
    """

    def __init__(self, group: Group, key: Pedersen, tag, gamma: int, *,
                 net: bool, rail: str = "", max_position: int = MAX_POSITION):
        self.group = group
        self.key = key
        self.tag = tag
        self.gamma = gamma
        self.tagged = tag.key_for(key)
        self.net = net
        self.rail = rail
        self.max_position = max_position
        self.positions: dict[bytes, Position] = {}
        self.credit: dict[bytes, CreditLine] = {}

    def open(self, handle: bytes, commitment) -> Position:
        if handle in self.positions:
            raise ValueError("position already open")
        self.positions[handle] = Position(handle, commitment, commitment)
        return self.positions[handle]

    def balance(self, handle: bytes):
        return self.positions[handle].commitment

    # --- how far below zero a position may go ----------------------------
    def grant(self, line: CreditLine) -> tuple[bool, str]:
        """Underwrite a net debit cap the infrastructure cannot see.

        The cap is committed, not published. That hides its size, and it also
        removes the sign: a coverage proof afterwards shows only that position
        plus cap is in range, never which side of zero the position was on.
        """
        if line.handle not in self.positions:
            return False, "unknown position handle"
        if line.rail != self.rail:
            return False, f"a cap for {line.rail!r} does not relax {self.rail!r}"
        # checked under the credit module's own domain, not the cycle's: a line
        # has to verify the same way outside the cycle that granted it, and
        # under the rail's own key, since that is what the cap is denominated in
        ok, reason = check_credit(self.tagged, line)
        if not ok:
            return False, reason
        self.credit[line.handle] = line
        return True, "ok"

    def _cap_commitment(self, handle: bytes):
        """A participant with no line has a cap of zero, which is the identity."""
        line = self.credit.get(handle)
        return line.cap_commitment if line else self.group.identity()

    def _headroom(self, commitment, handle: bytes):
        """position + cap, which is what has to be non-negative."""
        return self.group.mul(commitment, self._cap_commitment(handle))

    # --- building a move -------------------------------------------------
    def build_leg(self, payer: bytes, payee: bytes, amount: int, blinding: int,
                  position_value: int, position_blinding: int,
                  reference_commitment, reference_blinding: int,
                  context: bytes, cap_value: int = 0,
                  cap_blinding: int = 0) -> Leg:
        """Run by the payer, the only party that knows its position."""
        delta = self.tagged.commit(amount, blinding)
        delta_link = prove_same_value(
            self.key, self.tagged.g, self.key.g, delta, reference_commitment,
            amount, blinding, reference_blinding, context + b":link")
        cover = None
        if not self.net:
            headroom = position_value - amount + cap_value
            if headroom < 0:
                raise Rejected(
                    f"position {position_value} and a cap of {cap_value} cannot "
                    f"cover {amount}; on a gross rail the order is refused "
                    "before it exists")
            cover = prove_range(
                self.tagged, self.tagged.commit(headroom, (
                    position_blinding - blinding + cap_blinding) % self.group.order),
                headroom,
                (position_blinding - blinding + cap_blinding) % self.group.order,
                self.max_position.bit_length(), context + b":cover")
        return Leg(payer, payee, delta, delta_link, cover)

    # --- checking one -----------------------------------------------------
    def check(self, leg: Leg, reference_commitment,
              context: bytes) -> tuple[bool, str]:
        group = self.group
        for handle in (leg.payer, leg.payee):
            if handle not in self.positions:
                return False, "unknown position handle"
        if leg.payer == leg.payee:
            return False, "a leg cannot pay itself"
        if not verify_same_value(self.key, self.tagged.g, self.key.g, leg.delta,
                                 reference_commitment, leg.delta_link,
                                 context + b":link"):
            return False, "the amount leaving does not match the instruction"
        if self.net:
            if leg.cover is not None:
                return False, "a net rail must not carry a per-order cover proof"
            return True, "ok"
        if leg.cover is None:
            return False, "a gross rail needs a cover proof on every leg"
        residual = group.mul(self.positions[leg.payer].commitment,
                             group.neg(leg.delta))
        if not verify_range(self.tagged, self._headroom(residual, leg.payer),
                            leg.cover, context + b":cover"):
            return False, "the order would leave the position short of its cap"
        return True, "ok"

    def apply(self, leg: Leg) -> None:
        group = self.group
        payer, payee = self.positions[leg.payer], self.positions[leg.payee]
        payer.commitment = group.mul(payer.commitment, group.neg(leg.delta))
        payee.commitment = group.mul(payee.commitment, leg.delta)
        payer.moves += 1
        payee.moves += 1

    # --- the close --------------------------------------------------------
    def build_coverage(self, handle: bytes, value: int, blinding: int,
                       context: bytes, cap_value: int = 0,
                       cap_blinding: int = 0) -> Coverage:
        headroom = value + cap_value
        if headroom < 0:
            raise Rejected(
                f"net position {value} is short of its {cap_value} cap at the "
                "close; the cycle goes to the waterfall")
        combined_blinding = (blinding + cap_blinding) % self.group.order
        proof = prove_range(
            self.tagged, self.tagged.commit(headroom, combined_blinding),
            headroom, combined_blinding, self.max_position.bit_length(),
            context + b":close")
        return Coverage(handle, proof)

    def check_coverage(self, coverage: Coverage,
                       context: bytes) -> tuple[bool, str]:
        if coverage.handle not in self.positions:
            return False, "unknown position handle"
        headroom = self._headroom(self.positions[coverage.handle].commitment,
                                  coverage.handle)
        if not verify_range(self.tagged, headroom, coverage.proof,
                            context + b":close"):
            return False, "net position is not covered, even against its cap"
        return True, "ok"

    # --- what the cycle checks for free -----------------------------------
    def conserved(self) -> bool:
        """No proof and no opening: the two products simply have to agree."""
        group = self.group
        total = opened = group.identity()
        for position in self.positions.values():
            total = group.mul(total, position.commitment)
            opened = group.mul(opened, position.opening_commitment)
        return group.encode(total) == group.encode(opened)

    def snapshot(self) -> bytes:
        digest = hashlib.sha256(CYCLE_DOMAIN + (b"net" if self.net else b"gross"))
        for handle in sorted(self.positions):
            digest.update(len(handle).to_bytes(4, "big"))
            digest.update(handle)
            digest.update(self.group.encode(self.positions[handle].commitment))
        return digest.digest()


class NettingCycle:
    """A batch of orders under one mode, all-or-nothing at the close."""

    def __init__(self, group: Group, key: Pedersen, mode: NettingMode,
                 securities_tag, securities_gamma: int,
                 cash_tag, cash_gamma: int,
                 venue: SettlementVenue | None = None,
                 max_position: int = MAX_POSITION,
                 attest_batch: bool = False):
        self.attest_batch = attest_batch
        if attest_batch and mode is not NettingMode.NET_NET:
            raise ValueError(
                "a batch attestation replaces the per-trade instruction, which "
                "only makes sense when both rails are netted")
        self.group = group
        self.key = key
        self.mode = mode
        self.venue = venue or SettlementVenue(group, key)
        if group.encode(key.h) != group.encode(self.venue.key.h):
            raise ValueError(
                "the instruction venue and the position books use different "
                "commitment keys; bindings between them would be meaningless")
        self.securities = PositionBook(group, key, securities_tag, securities_gamma,
                                       net=mode.securities_net, rail="securities",
                                       max_position=max_position)
        self.cash = PositionBook(group, key, cash_tag, cash_gamma,
                                 net=mode.cash_net, rail="cash",
                                 max_position=max_position)
        self.admitted: list[Order] = []
        self.granted: list[CreditLine] = []
        self.refused: list[str] = []
        self.closed = False

    def admit(self, order: Order, *, now: int,
              context: bytes = CYCLE_DOMAIN) -> tuple[bool, str]:
        """Verify and apply, or refuse and change nothing."""
        if self.closed:
            return False, "the cycle is closed"
        if self.attest_batch:
            return self._accumulate(order, context)
        ok, reason = self.venue.verify(order.instruction, now=now)
        if not ok:
            self.refused.append(reason)
            return False, f"instruction: {reason}"

        instruction = order.instruction
        if not verify_product(self.key, instruction.price_commitment,
                              instruction.amount_commitment, order.cash_reference,
                              order.value_proof, context + b":value"):
            self.refused.append("cash is not quantity times price")
            return False, "cash is not quantity times price"

        checks = (
            (self.securities, order.securities, instruction.amount_commitment,
             context + b":sec"),
            (self.cash, order.cash, order.cash_reference, context + b":cash"),
        )
        for book, leg, commitment, ctx in checks:
            ok, reason = book.check(leg, commitment, ctx)
            if not ok:
                self.refused.append(reason)
                return False, reason
        # every check passed, so nothing can fail half way through
        for book, leg, _, _ in checks:
            book.apply(leg)
        spent, reason = self.venue.settle(instruction, now=now)
        if not spent:
            self.refused.append(reason)
            return False, reason
        self.admitted.append(order)
        return True, "ok"

    def admit_with_credit(self, order: Order, *, line: CreditLine, now: int,
                          pledge: Leg | None = None, pledge_reference=None,
                          context: bytes = CYCLE_DOMAIN) -> tuple[bool, str]:
        """Pledge, underwrite, draw and pay as one event.

        A participant that needs the overdraft to make the payment cannot do
        these in sequence: a pledge that lands without the limit being granted
        has locked collateral for nothing, and a limit granted without the
        pledge is unsecured. So everything is checked first and applied only if
        all of it holds --- the pledge on the collateral rail, the limit against
        that pledge, and the payment the limit exists to permit.
        """
        if self.closed:
            return False, "the cycle is closed"

        if pledge is not None:
            if pledge_reference is None:
                return False, "a pledge without the commitment it is stated against"
            ok, reason = self.securities.check(pledge, pledge_reference,
                                               context + b":pledge")
            if not ok:
                return False, f"pledge: {reason}"

        books = {"securities": self.securities, "cash": self.cash}
        if line.rail not in books:
            return False, f"credit: unknown rail {line.rail!r}"
        book = books[line.rail]
        ok, reason = check_credit(book.tagged, line)
        if not ok:
            return False, f"credit: {reason}"

        # the limit has to be in place for the payment to be admissible, so it is
        # staged rather than applied: if the payment fails, the grant unwinds
        previous = book.credit.get(line.handle)
        book.credit[line.handle] = line
        if pledge is not None:
            self.securities.apply(pledge)
        accepted, why = self.admit(order, now=now, context=context)
        if not accepted:
            if previous is None:
                book.credit.pop(line.handle, None)
            else:
                book.credit[line.handle] = previous
            if pledge is not None:
                self.securities.apply(
                    Leg(pledge.payee, pledge.payer, pledge.delta,
                        pledge.delta_link, None))
            return False, why
        self.granted.append(line)
        return True, "ok"

    def _accumulate(self, order: Order, context: bytes) -> tuple[bool, str]:
        """Apply a leg without verifying it, because the batch will be attested.

        Handles and self-payment are still checked: they cost nothing and a
        malformed cycle is not something an attestation should be able to
        authorise.
        """
        for book, leg in ((self.securities, order.securities), (self.cash, order.cash)):
            for handle in (leg.payer, leg.payee):
                if handle not in book.positions:
                    return False, "unknown position handle"
            if leg.payer == leg.payee:
                return False, "a leg cannot pay itself"
        for book, leg in ((self.securities, order.securities), (self.cash, order.cash)):
            book.apply(leg)
        self.admitted.append(order)
        return True, "ok"

    def batch_digest(self) -> bytes:
        """What an attestation signs: the closing positions and nothing else."""
        digest = hashlib.sha256(CYCLE_DOMAIN + b":batch:" + self.mode.value.encode())
        digest.update(self.securities.snapshot())
        digest.update(self.cash.snapshot())
        digest.update(len(self.admitted).to_bytes(4, "big"))
        return digest.digest()

    def close(self, coverages: dict[str, list[Coverage]] | None = None, *,
              now: int, context: bytes = CYCLE_DOMAIN,
              attestation: BatchAttestation | None = None) -> CycleReceipt:
        """Check every net rail is covered, then make the cycle final."""
        before_sec = self.securities.snapshot().hex()
        before_cash = self.cash.snapshot().hex()
        coverages = coverages or {}
        status, reason = "closed", "ok"

        if self.attest_batch:
            if attestation is None:
                status, reason = "failed", "an attested cycle needs an attestation"
            elif attestation.digest != self.batch_digest():
                status, reason = "failed", "the attestation is for other positions"

        for name, book in (("securities", self.securities), ("cash", self.cash)):
            if not book.net:
                continue
            supplied = {c.handle for c in coverages.get(name, ())}
            if supplied != set(book.positions):
                status, reason = "failed", f"{name}: not every net position is covered"
                break
            for coverage in coverages[name]:
                ok, why = book.check_coverage(coverage, context + b":" + name.encode())
                if not ok:
                    status, reason = "failed", f"{name}: {why}"
                    break
            if status == "failed":
                break

        if status == "closed":
            if not (self.securities.conserved() and self.cash.conserved()):
                status, reason = "failed", "a rail does not conserve value"

        if status == "closed":
            self.closed = True
        return CycleReceipt(
            mode=self.mode.value, admitted=len(self.admitted),
            refused=len(self.refused), status=status, reason=reason,
            securities_before=before_sec,
            securities_after=self.securities.snapshot().hex(),
            cash_before=before_cash, cash_after=self.cash.snapshot().hex(),
            closed_at=now)

File: rust/qomm-defmi/tests/netting.rs
//! Netting cycles, and the trade each mode makes.
//!
//! The behavioural distinction is admissibility: a gross rail refuses an order
//! it cannot cover now, a net rail waits until the close. The pair of tests
//! around that is the point of the file.

use curve25519_dalek::ristretto::RistrettoPoint;
use curve25519_dalek::scalar::Scalar;
use qomm_defmi::assets::AssetRegistry;
use qomm_defmi::credit::CreditCtx;
use qomm_defmi::netting::*;
use qomm_zk::pedersen::Pedersen;
use qomm_zkpi::{deal_quorum, frost, Bounds, Issuer, Venue};
use rand::rngs::OsRng;
use std::collections::BTreeMap;

const BITS: usize = 32;

struct Holder { securities: (i64, Scalar), cash: (i64, Scalar) }

struct Fixture {
    key: Pedersen,
    cycle: Cycle,
    issuer: Issuer,
    shares: BTreeMap<frost::Identifier, frost::keys::KeyPackage>,
    public: frost::keys::PublicKeyPackage,
    holders: BTreeMap<Vec<u8>, Holder>,
    credit: CreditCtx,
}

fn fixture(rng: &mut OsRng, mode: Mode, attest: bool, participants: usize,
           securities: &[(usize, i64)]) -> Fixture {
    let key = Pedersen::new(b"qomm:defmi:v1");
    let registry = AssetRegistry::new(key.clone(), 16);
    let (sec_tag, _) = registry.blind(3, false, rng).unwrap();
    let (cash_tag, _) = registry.blind(0, false, rng).unwrap();
    let mut sec_book = PositionBook::new(key.clone(), sec_tag, mode.securities_net(),
                                         "securities", BITS);
    let mut cash_book = PositionBook::new(key.clone(), cash_tag, mode.cash_net(), "cash", BITS);
    let mut holders = BTreeMap::new();
    for i in 0..participants {
        let handle = format!("p{i}").into_bytes();
        let opening = securities.iter().find(|(j, _)| *j == i).map(|(_, v)| *v).unwrap_or(10_000);
        let holder = Holder {
            securities: (opening, Scalar::random(rng)),
            cash: (10_000_000, Scalar::random(rng)),
        };
        sec_book.open(&handle, sec_book.tagged()
            .commit(&signed(holder.securities.0), &holder.securities.1));
        cash_book.open(&handle, cash_book.tagged()
            .commit(&signed(holder.cash.0), &holder.cash.1));
        holders.insert(handle, holder);
    }
    let (secret, public) = deal_quorum(7, 3, rng).unwrap();
    let shares = secret.into_iter()
        .map(|(id, s)| (id, frost::keys::KeyPackage::try_from(s).unwrap())).collect();
    let venue = Venue::new(key.clone(), &Bounds::default(), public.clone());
    Fixture {
        credit: CreditCtx::new(key.clone(), 64),
        issuer: Issuer::new(key.clone(), Bounds::default()),
        cycle: Cycle::new(key.clone(), mode, sec_book, cash_book, venue, attest).unwrap(),
        key, shares, public, holders,
    }
}

fn signed(value: i64) -> Scalar {
    if value >= 0 { Scalar::from(value as u64) } else { -Scalar::from((-value) as u64) }
}

fn sign(f: &Fixture, message: &[u8], rng: &mut OsRng) -> frost::Signature {
    let chosen: Vec<_> = f.shares.keys().take(3).cloned().collect();
    let mut nonces = BTreeMap::new();
    let mut commitments = BTreeMap::new();
    for id in &chosen {
        let (n, c) = frost::round1::commit(f.shares[id].signing_share(), rng);
        nonces.insert(*id, n);
        commitments.insert(*id, c);
    }
    let package = frost::SigningPackage::new(commitments, message);
    let mut sig_shares = BTreeMap::new();
    for id in &chosen {
        sig_shares.insert(*id, frost::round2::sign(&package, &nonces[id], &f.shares[id]).unwrap());
    }
    frost::aggregate(&package, &sig_shares, &f.public).unwrap()
}

/// Returns the order and the two delta blindings, which the counterparties have
/// to carry forward: a position is only usable by whoever knows its blinding.
#[allow(clippy::too_many_arguments)]
fn order(f: &Fixture, seller: &[u8], buyer: &[u8], quantity: u64, price: u64,
         nonce: u8, sec_cap: u64, sec_cap_blinding: Scalar, rng: &mut OsRng)
    -> Result<(Order, Scalar, Scalar), &'static str> {
    let (digest, openings, partial) = f.issuer.build(
        quantity, price, 3,
        RistrettoPoint::mul_base(&Scalar::from(11u64)),
        RistrettoPoint::mul_base(&Scalar::from(22u64)),
        1_500, [nonce; 32], 1_599_845, rng)?;
    let instruction = partial.sealed(sign(f, &digest, rng));
    let value = quantity * price;
    let cash_blinding = Scalar::random(rng);
    let cash_reference = f.key.commit_u64(value, &cash_blinding);
    let value_proof = prove_cash_reference(
        &f.key, &instruction.price_commitment, price, &openings.price,
        quantity, &openings.amount, &cash_blinding, rng);

    let sold = &f.holders[seller];
    let bought = &f.holders[buyer];
    let sec_delta = Scalar::random(rng);
    let cash_delta = Scalar::random(rng);
    let sec_leg = f.cycle.securities.build_leg(
        seller, buyer, quantity, &sec_delta,
        sold.securities.0, &sold.securities.1,
        &instruction.amount_commitment, &openings.amount,
        sec_cap, &sec_cap_blinding, rng)?;
    let cash_leg = f.cycle.cash.build_leg(
        buyer, seller, value, &cash_delta,
        bought.cash.0, &bought.cash.1,
        &cash_reference, &cash_blinding, 0, &Scalar::ZERO, rng)?;
    Ok((Order { instruction, securities: sec_leg, cash: cash_leg, cash_reference, value_proof },
        sec_delta, cash_delta))
}

/// Both sides update their own books, which the cycle never sees.
fn settle_books(f: &mut Fixture, seller: &[u8], buyer: &[u8], quantity: u64, value: u64,
                sec_delta: Scalar, cash_delta: Scalar) {
    {
        let s = f.holders.get_mut(&seller.to_vec()).unwrap();
        s.securities.0 -= quantity as i64;
        s.securities.1 -= sec_delta;
        s.cash.0 += value as i64;
        s.cash.1 += cash_delta;
    }
    let b = f.holders.get_mut(&buyer.to_vec()).unwrap();
    b.securities.0 += quantity as i64;
    b.securities.1 += sec_delta;
    b.cash.0 -= value as i64;
    b.cash.1 -= cash_delta;
}

#[test]
fn every_mode_settles_and_conserves() {
    let mut rng = OsRng;
    for (mode, attest) in [(Mode::GrossGross, false), (Mode::GrossNet, false),
                           (Mode::NetNet, false), (Mode::NetNet, true)] {
        let mut f = fixture(&mut rng, mode, attest, 4, &[]);
        for nonce in 0..4u8 {
            let (o, sd, cd) = order(&f, b"p0", b"p1", 40, 10_000, nonce, 0,
                                    Scalar::ZERO, &mut rng).expect("build");
            f.cycle.admit(&o, 1_000, &mut rng).expect("admit");
            settle_books(&mut f, b"p0", b"p1", 40, 40 * 10_000, sd, cd);
        }
        assert!(f.cycle.securities.conserved() && f.cycle.cash.conserved(),
                "{mode:?} attest={attest}");
    }
}

#[test]
fn a_gross_rail_refuses_a_delivery_it_cannot_cover_yet() {
    // settlement cannot fail, but the order cannot exist
    let mut rng = OsRng;
    let f = fixture(&mut rng, Mode::GrossGross, false, 3, &[(0, 0)]);
    assert_eq!(
        order(&f, b"p0", b"p2", 100, 10_000, 1, 0, Scalar::ZERO, &mut rng).err(),
        Some("the order would leave the position short of its cap"));
}

#[test]
fn a_net_rail_lets_the_offsetting_pair_through_in_either_order() {
    // and this is what the trade buys back: order-insensitivity
    let mut rng = OsRng;
    for delivery_first in [true, false] {
        let mut f = fixture(&mut rng, Mode::NetNet, false, 3, &[(0, 0)]);
        let mut pairs: Vec<(&[u8], &[u8])> = vec![(b"p0", b"p2"), (b"p1", b"p0")];
        if !delivery_first { pairs.reverse(); }
        for (nonce, (seller, buyer)) in pairs.into_iter().enumerate() {
            let (o, sd, cd) = order(&f, seller, buyer, 100, 10_000, nonce as u8, 0,
                                    Scalar::ZERO, &mut rng).expect("build");
            f.cycle.admit(&o, 1_000, &mut rng).expect("admit");
            settle_books(&mut f, seller, buyer, 100, 100 * 10_000, sd, cd);
        }
        let coverage: Vec<_> = f.holders.iter()
            .map(|(h, holder)| f.cycle.securities
                .build_coverage(h, holder.securities.0, &holder.securities.1, 0, &Scalar::ZERO)
                .expect("coverage"))
            .collect();
        let cash: Vec<_> = f.holders.iter()
            .map(|(h, holder)| f.cycle.cash
                .build_coverage(h, holder.cash.0, &holder.cash.1, 0, &Scalar::ZERO)
                .expect("cash coverage"))
            .collect();
        assert_eq!(f.cycle.close(&coverage, &cash, None), Ok(()));
    }
}

#[test]
fn a_cap_lets_a_net_position_go_below_zero_and_no_further() {
    let mut rng = OsRng;
    let mut f = fixture(&mut rng, Mode::NetNet, false, 3, &[(0, 0)]);
    let cap = 300u64;
    let cap_blinding = Scalar::random(&mut rng);
    // granted under the rail's own key: a cap in other units is incomparable
    let rail_credit = CreditCtx::new(f.cycle.securities.tagged(), 64);
    let line = rail_credit.grant(b"p0", "securities", cap, &cap_blinding,
                                 10_000, &Scalar::random(&mut rng), 500).unwrap();
    f.cycle.securities.grant(&rail_credit, line).expect("grant");

    // the buyer pays out of a finite cash position, so the trade has to fit it
    let (o, sd, cd) = order(&f, b"p0", b"p2", 200, 10_000, 1, 0, Scalar::ZERO,
                            &mut rng).unwrap();
    f.cycle.admit(&o, 1_000, &mut rng).expect("admit");
    settle_books(&mut f, b"p0", b"p2", 200, 200 * 10_000, sd, cd);
    assert_eq!(f.holders[&b"p0".to_vec()].securities.0, -200);

    let coverage: Vec<_> = f.holders.iter()
        .map(|(h, holder)| {
            let (c, cb) = if h == b"p0" { (cap, cap_blinding) } else { (0, Scalar::ZERO) };
            f.cycle.securities
                .build_coverage(h, holder.securities.0, &holder.securities.1, c, &cb)
                .expect("coverage")
        })
        .collect();
    let cash: Vec<_> = f.holders.iter()
        .map(|(h, holder)| f.cycle.cash
            .build_coverage(h, holder.cash.0, &holder.cash.1, 0, &Scalar::ZERO).unwrap())
        .collect();
    assert_eq!(f.cycle.close(&coverage, &cash, None), Ok(()));
}

#[test]
fn a_position_beyond_its_cap_fails_at_the_close() {
    let mut rng = OsRng;
    let mut f = fixture(&mut rng, Mode::NetNet, false, 3, &[(0, 0)]);
    let cap_blinding = Scalar::random(&mut rng);
    let rail_credit = CreditCtx::new(f.cycle.securities.tagged(), 64);
    let line = rail_credit.grant(b"p0", "securities", 300, &cap_blinding,
                                 10_000, &Scalar::random(&mut rng), 500).unwrap();
    f.cycle.securities.grant(&rail_credit, line).unwrap();
    let (o, _, _) = order(&f, b"p0", b"p2", 400, 10_000, 1, 0, Scalar::ZERO,
                          &mut rng).unwrap();
    f.cycle.admit(&o, 1_000, &mut rng).unwrap();
    assert!(f.cycle.securities
        .build_coverage(b"p0", -400, &f.holders[&b"p0".to_vec()].securities.1,
                        300, &cap_blinding)
        .is_err());
}

#[test]
fn a_leg_cannot_pay_itself() {
    let mut rng = OsRng;
    let mut f = fixture(&mut rng, Mode::NetNet, false, 3, &[]);
    let (mut o, _, _) = order(&f, b"p0", b"p1", 40, 100_000, 1, 0, Scalar::ZERO,
                              &mut rng).unwrap();
    o.securities.payee = o.securities.payer.clone();
    assert_eq!(f.cycle.admit(&o, 1_000, &mut rng), Err("a leg cannot pay itself"));
}

#[test]
fn an_instruction_settles_once_per_cycle() {
    let mut rng = OsRng;
    let mut f = fixture(&mut rng, Mode::GrossGross, false, 3, &[]);
    let (o, _, _) = order(&f, b"p0", b"p1", 40, 100_000, 1, 0, Scalar::ZERO,
                          &mut rng).unwrap();
    assert_eq!(f.cycle.admit(&o, 1_000, &mut rng), Ok(()));
    assert_eq!(f.cycle.admit(&o, 1_000, &mut rng), Err("already settled"));
}

#[test]
fn a_batch_attestation_only_makes_sense_when_both_rails_net() {
    let mut rng = OsRng;
    let key = Pedersen::new(b"qomm:defmi:v1");
    let registry = AssetRegistry::new(key.clone(), 16);
    let (tag, _) = registry.blind(3, false, &mut rng).unwrap();
    let (secret, public) = deal_quorum(7, 3, &mut rng).unwrap();
    drop(secret);
    for mode in [Mode::GrossGross, Mode::GrossNet] {
        let books = || PositionBook::new(key.clone(), tag.clone(), false, "securities", BITS);
        let venue = Venue::new(key.clone(), &Bounds::default(), public.clone());
        assert!(Cycle::new(key.clone(), mode, books(), books(), venue, true).is_err());
    }
}

#[test]
fn an_attestation_for_other_positions_is_refused() {
    let mut rng = OsRng;
    let mut f = fixture(&mut rng, Mode::NetNet, true, 3, &[]);
    let (o, sd, cd) = order(&f, b"p0", b"p1", 40, 10_000, 1, 0, Scalar::ZERO,
                            &mut rng).unwrap();
    f.cycle.admit(&o, 1_000, &mut rng).unwrap();
    settle_books(&mut f, b"p0", b"p1", 40, 40 * 10_000, sd, cd);
    let coverage: Vec<_> = f.holders.iter()
        .map(|(h, holder)| f.cycle.securities
            .build_coverage(h, holder.securities.0, &holder.securities.1, 0, &Scalar::ZERO))
        .filter_map(|c| c.ok())
        .collect();
    assert_eq!(
        f.cycle.close(&coverage, &[], Some(&BatchAttestation { digest: [0u8; 32] })),
        Err("the attestation is for other positions"));
}
